From 12c3d2e72273cddb0867bc1d78094cee0f8c43cb Mon Sep 17 00:00:00 2001
From: Aniket Randive <quic_arandive@quicinc.com>
Date: Mon, 7 Apr 2025 14:59:37 +0530
Subject: [PATCH 08/12] i2c: i2c-qcom-geni: Update counter setting for CCU QUP

Add 38.4MHz serial engine clock in Titan clock controller
to achieve 1MHz bus frequency. Include counter settings
for 38.4MHz input serial engine clock. Use the same counter
settings to achieve 100KHz or 400KHz with a 38.4MHz source
clock as used with a 19.2MHz source clock.

Introduce a separate counter table for CCU I2C serial engine.

Change-Id: I1cf3a0c0f2465712024af19de03d5706d323677a
Signed-off-by: Aniket Randive <quic_arandive@quicinc.com>
---
 drivers/i2c/busses/i2c-qcom-geni.c | 38 +++++++++++++++++++++++++-----
 1 file changed, 32 insertions(+), 6 deletions(-)

diff --git a/drivers/i2c/busses/i2c-qcom-geni.c b/drivers/i2c/busses/i2c-qcom-geni.c
index 3a7b0871fd47..3a1208672939 100644
--- a/drivers/i2c/busses/i2c-qcom-geni.c
+++ b/drivers/i2c/busses/i2c-qcom-geni.c
@@ -288,6 +288,18 @@ static const struct geni_i2c_clk_fld geni_i2c_clk_map[] = {
 	{KHZ(1000), 1, 3,  9, 18},
 };
 
+/*
+ * For CCU QUP, to achieve 1MHz bus frequency, in addition to
+ * existing 19.2MHz serial engine clock, 38.4MHz serial engine clock
+ * is added in Titan clock controller. This programming is applicable
+ * only for CCU QUP.
+ */
+static const struct geni_i2c_clk_fld geni_ccu_i2c_clk_map[] = {
+	{KHZ(100), 7, 10, 11, 26},
+	{KHZ(400), 2, 5, 12, 24},
+	{KHZ(1000), 1, 7, 20, 36},
+};
+
 static int setup_gpi_dma(struct geni_i2c_dev *gi2c);
 static void release_gpi_dma(struct geni_i2c_dev *gi2c);
 static void i2c_gpi_cb_result(void *cb, const struct dmaengine_result *result);
@@ -316,12 +328,17 @@ static DEVICE_ATTR_RW(log_level);
 
 static int geni_i2c_clk_map_idx(struct geni_i2c_dev *gi2c)
 {
-	int i;
-	const struct geni_i2c_clk_fld *itr = geni_i2c_clk_map;
+	unsigned int i;
+	const struct geni_i2c_clk_fld *itr;
+	size_t num_elements;
+
+	itr = gi2c->is_ccu_qup ? geni_ccu_i2c_clk_map : geni_i2c_clk_map;
+	num_elements = gi2c->is_ccu_qup ? ARRAY_SIZE(geni_ccu_i2c_clk_map)
+			: ARRAY_SIZE(geni_i2c_clk_map);
 
-	for (i = 0; i < ARRAY_SIZE(geni_i2c_clk_map); i++, itr++) {
-		if (itr->clk_freq_out == gi2c->clk_freq_out) {
-			gi2c->clk_fld = itr;
+	for (i = 0; i < num_elements; i++) {
+		if (itr[i].clk_freq_out == gi2c->clk_freq_out) {
+			gi2c->clk_fld = &itr[i];
 			return 0;
 		}
 	}
@@ -333,7 +350,16 @@ static void qcom_geni_i2c_conf(struct geni_i2c_dev *gi2c)
 	const struct geni_i2c_clk_fld *itr = gi2c->clk_fld;
 	u32 val;
 
-	writel_relaxed(0, gi2c->se.base + SE_GENI_CLK_SEL);
+	/*
+	 * For CCU QUP, to achieve 1MHz bus frequency
+	 * 38.4MHz serial engine clock is needed. otherwise
+	 * 19.2MHz serial engine clock need to be use for 100K
+	 * or 400K frequency.
+	 */
+	if (gi2c->is_ccu_qup && (itr->clk_freq_out == KHZ(1000)))
+		writel_relaxed(1, gi2c->se.base + SE_GENI_CLK_SEL);
+	else
+		writel_relaxed(0, gi2c->se.base + SE_GENI_CLK_SEL);
 
 	val = (itr->clk_div << CLK_DIV_SHFT) | SER_CLK_EN;
 	writel_relaxed(val, gi2c->se.base + GENI_SER_M_CLK_CFG);
-- 
2.34.1

